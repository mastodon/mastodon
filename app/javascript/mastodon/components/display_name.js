import React from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import PropTypes from 'prop-types';
import { autoPlayGif } from 'mastodon/initial_state';
import Skeleton from 'mastodon/components/skeleton';

export default class DisplayName extends React.PureComponent {

  static propTypes = {
    account: ImmutablePropTypes.map,
    others: ImmutablePropTypes.list,
    localDomain: PropTypes.string,
  };

  handleMouseEnter = ({ currentTarget }) => {
    if (autoPlayGif) {
      return;
    }

    const emojis = currentTarget.querySelectorAll('.custom-emoji');

    for (var i = 0; i < emojis.length; i++) {
      let emoji = emojis[i];
      emoji.src = emoji.getAttribute('data-original');
    }
  };

  handleMouseLeave = ({ currentTarget }) => {
    if (autoPlayGif) {
      return;
    }

    const emojis = currentTarget.querySelectorAll('.custom-emoji');

    for (var i = 0; i < emojis.length; i++) {
      let emoji = emojis[i];
      emoji.src = emoji.getAttribute('data-static');
    }
  };

  render () {
    const { others, localDomain } = this.props;

    let displayName, suffix, account;

    if (others && others.size > 1) {
      displayName = others.take(2).map(a => <bdi key={a.get('id')}><strong className='display-name__html' dangerouslySetInnerHTML={{ __html: a.get('display_name_html') }} /></bdi>).reduce((prev, cur) => [prev, ', ', cur]);

      if (others.size - 2 > 0) {
        suffix = `+${others.size - 2}`;
      }
    } else if ((others && others.size > 0) || this.props.account) {
      if (others && others.size > 0) {
        account = others.first();
      } else {
        account = this.props.account;
      }

      let acct = account.get('acct');

      if (acct.indexOf('@') === -1 && localDomain) {
        acct = `${acct}@${localDomain}`;
      }

      displayName = <bdi><strong className='display-name__html' dangerouslySetInnerHTML={{ __html: account.get('display_name_html') }} /></bdi>;
      suffix      = <span className='display-name__account'>@{acct}</span>;
    } else {
      displayName = <bdi><strong className='display-name__html'><Skeleton width='10ch' /></strong></bdi>;
      suffix = <span className='display-name__account'><Skeleton width='7ch' /></span>;
    }

    return (
      <span className='display-name' onMouseEnter={this.handleMouseEnter} onMouseLeave={this.handleMouseLeave}>
        {displayName} {suffix}
      </span>
    );
  }

}
