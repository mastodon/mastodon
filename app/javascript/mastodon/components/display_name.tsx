import React from 'react';

import type { List } from 'immutable';

import type { Account } from 'mastodon/models/account';

import { autoPlayGif } from '../initial_state';

import { Skeleton } from './skeleton';

interface Props {
  account?: Account;
  others?: List<Account>;
  localDomain?: string;
}

export class DisplayName extends React.PureComponent<Props> {
  handleMouseEnter: React.ReactEventHandler<HTMLSpanElement> = ({
    currentTarget,
  }) => {
    if (autoPlayGif) {
      return;
    }

    const emojis =
      currentTarget.querySelectorAll<HTMLImageElement>('img.custom-emoji');

    emojis.forEach((emoji) => {
      const originalSrc = emoji.getAttribute('data-original');
      if (originalSrc != null) emoji.src = originalSrc;
    });
  };

  handleMouseLeave: React.ReactEventHandler<HTMLSpanElement> = ({
    currentTarget,
  }) => {
    if (autoPlayGif) {
      return;
    }

    const emojis =
      currentTarget.querySelectorAll<HTMLImageElement>('img.custom-emoji');

    emojis.forEach((emoji) => {
      const staticSrc = emoji.getAttribute('data-static');
      if (staticSrc != null) emoji.src = staticSrc;
    });
  };

  render() {
    const { others, localDomain } = this.props;

    let displayName: React.ReactNode,
      suffix: React.ReactNode,
      account: Account | undefined;

    if (others && others.size > 0) {
      account = others.first();
    } else if (this.props.account) {
      account = this.props.account;
    }

    if (others && others.size > 1) {
      displayName = others
        .take(2)
        .map((a) => (
          <bdi key={a.get('id')}>
            <strong
              className='display-name__html'
              dangerouslySetInnerHTML={{ __html: a.get('display_name_html') }}
            />
          </bdi>
        ))
        .reduce((prev, cur) => [prev, ', ', cur]);

      if (others.size - 2 > 0) {
        suffix = `+${others.size - 2}`;
      }
    } else if (account) {
      let acct = account.get('acct');

      if (!acct.includes('@') && localDomain) {
        acct = `${acct}@${localDomain}`;
      }

      displayName = (
        <bdi>
          <strong
            className='display-name__html'
            dangerouslySetInnerHTML={{
              __html: account.get('display_name_html'),
            }}
          />
        </bdi>
      );
      suffix = <span className='display-name__account'>@{acct}</span>;
    } else {
      displayName = (
        <bdi>
          <strong className='display-name__html'>
            <Skeleton width='10ch' />
          </strong>
        </bdi>
      );
      suffix = (
        <span className='display-name__account'>
          <Skeleton width='7ch' />
        </span>
      );
    }

    return (
      <span
        className='display-name'
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
      >
        {displayName} {suffix}
      </span>
    );
  }
}
