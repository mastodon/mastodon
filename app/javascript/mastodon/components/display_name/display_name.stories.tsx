import type { ComponentProps } from 'react';

import type { Meta, StoryObj } from '@storybook/react-vite';

import { accountFactoryState } from '@/testing/factories';

import { DisplayName, LinkedDisplayName } from './index';

type PageProps = Omit<ComponentProps<typeof DisplayName>, 'account'> & {
  name: string;
  username: string;
  loading: boolean;
};

const meta = {
  title: 'Components/DisplayName',
  args: {
    username: 'mastodon@mastodon.social',
    name: 'Test User ðŸ§ª',
    loading: false,
    simple: false,
    noDomain: false,
    localDomain: 'mastodon.social',
  },
  tags: [],
  render({ name, username, loading, ...args }) {
    const account = !loading
      ? accountFactoryState({
          display_name: name,
          acct: username,
        })
      : undefined;
    return <DisplayName {...args} account={account} />;
  },
} satisfies Meta<PageProps>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {},
};

export const Loading: Story = {
  args: {
    loading: true,
  },
};

export const NoDomain: Story = {
  args: {
    noDomain: true,
  },
};

export const Simple: Story = {
  args: {
    simple: true,
  },
};

export const LocalUser: Story = {
  args: {
    username: 'localuser',
    name: 'Local User',
    localDomain: '',
  },
};

export const Linked: Story = {
  render({ name, username, loading, ...args }) {
    const account = !loading
      ? accountFactoryState({
          display_name: name,
          acct: username,
        })
      : undefined;
    return <LinkedDisplayName {...args} account={account} />;
  },
};
