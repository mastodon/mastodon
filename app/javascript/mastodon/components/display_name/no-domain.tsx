import type { ComponentPropsWithoutRef, FC } from 'react';

import classNames from 'classnames';

import { AnimateEmojiProvider } from '../emoji/context';
import { EmojiHTML } from '../emoji/html';
import { Skeleton } from '../skeleton';

import type { DisplayNameProps } from './index';

export const DisplayNameWithoutDomain: FC<
  Omit<DisplayNameProps, 'variant' | 'localDomain'> &
    ComponentPropsWithoutRef<'span'>
> = ({ account, className, children, ...props }) => {
  return (
    <AnimateEmojiProvider
      {...props}
      as='span'
      className={classNames('display-name', className)}
    >
      <bdi>
        {account ? (
          <EmojiHTML
            className='display-name__html'
            htmlString={account.get('display_name_html')}
            as='strong'
            extraEmojis={account.get('emojis')}
          />
        ) : (
          <strong className='display-name__html'>
            <Skeleton width='10ch' />
          </strong>
        )}
      </bdi>
      {children}
    </AnimateEmojiProvider>
  );
};
