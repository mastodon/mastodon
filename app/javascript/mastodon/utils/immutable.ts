import { isImmutable } from 'immutable';

/**
 * Converts an immutable Map or Set to JS if it isn't already a plain JS object.
 * This is useful in code that needs to be able to deal with either type.
 */

export function immutableToJSIfNeeded<T>(object: T): T {
  return isImmutable(object) ? (object.toJS() as T) : object;
}
