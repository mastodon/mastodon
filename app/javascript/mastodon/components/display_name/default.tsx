import { useMemo } from 'react';
import type { ComponentPropsWithoutRef, FC } from 'react';

import { Skeleton } from '../skeleton';

import type { DisplayNameProps } from './index';
import { DisplayNameWithoutDomain } from './no-domain';

export const DisplayNameDefault: FC<
  Omit<DisplayNameProps, 'variant'> & ComponentPropsWithoutRef<'span'>
> = ({ account, localDomain, className, ...props }) => {
  const username = useMemo(() => {
    if (!account) {
      return null;
    }
    let acct = account.get('acct');

    if (!acct.includes('@') && localDomain) {
      acct = `${acct}@${localDomain}`;
    }
    return `@${acct}`;
  }, [account, localDomain]);

  return (
    <DisplayNameWithoutDomain
      account={account}
      className={className}
      {...props}
    >
      {' '}
      <span className='display-name__account'>
        {username ?? <Skeleton width='7ch' />}
      </span>
    </DisplayNameWithoutDomain>
  );
};
