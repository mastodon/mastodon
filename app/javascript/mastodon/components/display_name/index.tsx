import type { ComponentPropsWithoutRef, FC } from 'react';
import { useMemo } from 'react';

import classNames from 'classnames';
import type { LinkProps } from 'react-router-dom';
import { Link } from 'react-router-dom';

import { EmojiHTML } from '@/mastodon/features/emoji/emoji_html';
import type { Account } from '@/mastodon/models/account';
import { isModernEmojiEnabled } from '@/mastodon/utils/environment';

import { Skeleton } from '../skeleton';

interface Props {
  account?: Account;
  localDomain?: string;
  simple?: boolean;
  noDomain?: boolean;
}

export const DisplayName: FC<Props & ComponentPropsWithoutRef<'span'>> = ({
  account,
  localDomain,
  simple = false,
  noDomain = false,
  className,
  ...props
}) => {
  const username = useMemo(() => {
    if (!account || noDomain) {
      return null;
    }
    let acct = account.get('acct');

    if (!acct.includes('@') && localDomain) {
      acct = `${acct}@${localDomain}`;
    }
    return `@${acct}`;
  }, [account, localDomain, noDomain]);

  if (!account) {
    if (simple) {
      return null;
    }
    return (
      <span {...props} className={classNames('display-name', className)}>
        <bdi>
          <strong className='display-name__html'>
            <Skeleton width='10ch' />
          </strong>
        </bdi>
        {!noDomain && (
          <span className='display-name__account'>
            &nbsp;
            <Skeleton width='7ch' />
          </span>
        )}
      </span>
    );
  }
  const accountName = isModernEmojiEnabled()
    ? account.get('display_name')
    : account.get('display_name_html');
  if (simple) {
    return (
      <bdi>
        <EmojiHTML {...props} htmlString={accountName} shallow as='span' />
      </bdi>
    );
  }

  return (
    <span {...props} className={classNames('display-name', className)}>
      <bdi>
        <EmojiHTML
          className='display-name__html'
          htmlString={accountName}
          shallow
          as='strong'
        />
      </bdi>
      {username && (
        <span className='display-name__account'>&nbsp;{username}</span>
      )}
    </span>
  );
};

export const LinkedDisplayName: FC<
  Props & { asProps?: ComponentPropsWithoutRef<'span'> } & Partial<LinkProps>
> = ({
  account,
  asProps = {},
  className,
  localDomain,
  simple,
  noDomain,
  ...linkProps
}) => {
  const displayProps = {
    account,
    className,
    localDomain,
    simple,
    noDomain,
    ...asProps,
  };
  if (!account) {
    return <DisplayName {...displayProps} />;
  }

  return (
    <Link
      to={`/@${account.acct}`}
      title={`@${account.acct}`}
      data-hover-card-account={account.id}
      {...linkProps}
    >
      <DisplayName {...displayProps} />
    </Link>
  );
};
