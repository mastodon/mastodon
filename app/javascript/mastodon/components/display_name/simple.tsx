import type { ComponentPropsWithoutRef, FC } from 'react';

import { isModernEmojiEnabled } from '@/mastodon/utils/environment';

import { EmojiHTML } from '../emoji/html';

import type { DisplayNameProps } from './index';

export const DisplayNameSimple: FC<
  Omit<DisplayNameProps, 'variant' | 'localDomain'> &
    ComponentPropsWithoutRef<'span'>
> = ({ account, ...props }) => {
  if (!account) {
    return null;
  }

  return (
    <bdi>
      <EmojiHTML
        {...props}
        as='span'
        htmlString={
          isModernEmojiEnabled()
            ? account.get('display_name')
            : account.get('display_name_html')
        }
        extraEmojis={account.get('emojis')}
      />
    </bdi>
  );
};
