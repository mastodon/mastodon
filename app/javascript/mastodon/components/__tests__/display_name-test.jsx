import { fromJS }  from 'immutable';

import renderer from 'react-test-renderer';

import { DisplayName } from '../display_name';

describe('<DisplayName />', () => {
  it('renders display name + account name', () => {
    const account = fromJS({
      username: 'bar',
      acct: 'bar@baz',
      display_name_html: '<p>Foo</p>',
    });
    const component = renderer.create(<DisplayName account={account} />);
    const tree      = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
