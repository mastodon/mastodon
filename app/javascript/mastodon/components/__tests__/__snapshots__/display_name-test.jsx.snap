// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DisplayName /> renders display name + account name 1`] = `
<span
  className="display-name"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <bdi>
    <strong
      className="display-name__html"
      dangerouslySetInnerHTML={
        {
          "__html": "<p>Foo</p>",
        }
      }
    />
  </bdi>
   
  <span
    className="display-name__account"
  >
    @
    bar@baz
  </span>
</span>
`;
